### PR TITLE
doc: add note about checking pending stray directory entries before initiating upgrade

### PR DESCRIPTION
**Pull request by vshankar**
Signed-off-by: Venky Shankar <vshankar@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
@huww98 FYI

**Comment by huww98** [comment]
Stray files should be OK. Only stray directories would crash MDS. Getting rid of all stray files may be very hard, we always have thousands of them.

We use the following script to identify stray directories:

Dump the content of all strays (sleep to avoid overloading the MDS if we have a lot of strays):
```bash
for i in {0..9}; do ceph tell mds.<rank> dump tree "~mdsdir/stray${i}" > mds<rank>/stray${i}.json; sleep 2; done
```

Then run the following python script:
```Python
import json

for i_mds in range(2):
    for i_stray in range(10):
        l = json.load(open(f'mds{i_mds}/stray{i_stray}.json'))
        for inode in l:
            if inode['nlink'] > 0:
                continue
            if inode['dir_layout']['dir_hash'] == 0:
                continue
            print(f'mds{i_mds} ino:{inode["ino"]:x}', inode['stray_prior_path'], [c['client_id'] for c in inode['client_caps']])
```

**Comment by vshankar** [comment]
> Stray files should be OK. Only stray directories would crash MDS. Getting rid of all stray files may be very hard, we always have thousands of them.

Right. Ideally we would want the admins to check this via a single command (w/ jq magic) rather than running custom script. I'll see what I can come up with.

> 
> We use the following script to identify stray directories:
> 
> Dump the content of all strays (sleep to avoid overloading the MDS if we have a lot of strays):
> 
> ```shell
> for i in {0..9}; do ceph tell mds.<rank> dump tree "~mdsdir/stray${i}" > mds<rank>/stray${i}.json; sleep 2; done
> ```
> 
> Then run the following python script:
> 
> ```python
> import json
> 
> for i_mds in range(2):
>     for i_stray in range(10):
>         l = json.load(open(f'mds{i_mds}/stray{i_stray}.json'))
>         for inode in l:
>             if inode['nlink'] > 0:
>                 continue
>             if inode['dir_layout']['dir_hash'] == 0:
>                 continue
>             print(f'mds{i_mds} ino:{inode["ino"]:x}', inode['stray_prior_path'], [c['client_id'] for c in inode['client_caps']])
> ```

**Comment by vshankar** [comment]
@huww98 How about something like (e.g., for rank1, stray0):

```
ceph tell mds.1 dump tree ~mdsdir/stray0 2>/dev/null |  jq '.[] | select ((.nlink == 0 or .dir_layout.dir_hash > 0) and .stray_prior_path != "") | .stray_prior_path' | wc -l
```

**Comment by huww98** [comment]
> @huww98 How about something like (e.g., for rank1, stray0):
> 
> ```
> ceph tell mds.1 dump tree ~mdsdir/stray0 2>/dev/null |  jq '.[] | select ((.nlink == 0 or .dir_layout.dir_hash > 0) and .stray_prior_path != "") | .stray_prior_path' | wc -l
> ```

@vshankar should be `.nlink == 0 and .dir_layout.dir_hash > 0` (`and` instead of `or`)

Besides, why `.stray_prior_path != ""`? if it is really empty for some reason, we may need to identify the stray dir by other means, instead of ignoring it.

**Comment by vshankar** [comment]
> > @huww98 How about something like (e.g., for rank1, stray0):
> > ```
> > ceph tell mds.1 dump tree ~mdsdir/stray0 2>/dev/null |  jq '.[] | select ((.nlink == 0 or .dir_layout.dir_hash > 0) and .stray_prior_path != "") | .stray_prior_path' | wc -l
> > ```
> 
> @vshankar should be `.nlink == 0 and .dir_layout.dir_hash > 0` (`and` instead of `or`)
> 
> Besides, why `.stray_prior_path != ""`? if it is really empty for some reason, we may need to identify the stray dir by other means, instead of ignoring it.

ah yes! thx for noticing that.

regarding empty stray_prior_path, that's in the inode dump of the stray directory itself

```
    {
        "ino": 1541,
        ....
        ....
        "stray_prior_path": "",
        ....
        ....
    }
```

**Comment by huww98** [comment]
> regarding empty stray_prior_path, that's in the inode dump of the stray directory itself

That's why I use `.nlink == 0` in the filter. The stray directory itself should have nlink == 1

**Comment by vshankar** [comment]
> > regarding empty stray_prior_path, that's in the inode dump of the stray directory itself
> 
> That's why I use `.nlink == 0` in the filter. The stray directory itself should have nlink == 1

In that case, I think the inode dump is a bit incorrect for some of the dirs. Let me check.

**Comment by vshankar** [comment]
@huww98 OK -- this is what we need:

```
 jq '.[] | select (.nlink == 0 and .dir_layout.dir_hash > 0) | .stray_prior_path' | wc -l
```

jq was throwing an extra "\n" at the end that bumped up the count.

**Comment by vshankar** [comment]
updated and pushed - please take a look.

File: doc/releases/pacific.rst [hunk 1]
```diff
@@ -974,6 +974,19 @@ recovering OSDs).  (This is optional, but recommended.)
...
   https://tracker.ceph.com/issues/53062.
 
+.. note::
+   When using multiple active Ceph Metadata Servers, ensure that there are
+   no pending stray entries which are directories for active ranks except rank 0 as
+   starting an upgrade (which sets `max_mds` to 1) could crash the Ceph
+   Metadata Server. The following command should return zero (0) stray entries
+   for all stray directories::
+
+     # for idx in {0..9}; do ceph tell mds.<rank> dump tree ~mdsdir/stray$idx| jq '.[] | select (.nlink == 0and .dir_layout.dir_hash > 0) | .stray_prior_path' | wc -l; done
```
Comment (huww98) [inline doc/releases/pacific.rst:984]: Missing a space before `and`.

**Review by huww98 — CHANGES_REQUESTED** [review]
Missing a space before `and`.

**Comment by vshankar** [comment]
> Missing a space before `and`.

fixed.

**Review by huww98 — APPROVED** [review]

**Review by kotreshhr — APPROVED** [review]

**Review by lxbsz — APPROVED** [review]
LGTM.

**Pull request merged**